test/extra/test_utils.py

```diff
@@ -36,25 +36,33 @@ def __init__(self):
           d.as_strided([2, 2], [2, 3], storage_offset=4)
         )
 
-    for isfloat16 in [True, False]:
-      model = torch.nn.Sequential(
-        torch.nn.Linear(4, 8),
-        torch.nn.Linear(8, 3),
-        LayerWithOffset()
-      )
-      if isfloat16: model = model.half()
-
-      with tempfile.TemporaryDirectory() as tmpdirname:
-        path = tmpdirname + '/testloadmodel.pth'
-        torch.save(model.state_dict(), path)
-        model2 = fake_torch_load_zipped(path)
-
-      for name, a in model.state_dict().items():
-        b = model2[name]
-        a, b = a.numpy(), b.numpy()
-        assert a.shape == b.shape
-        assert a.dtype == b.dtype
-        assert np.array_equal(a, b)
+    for isfloat16_aftersave in [True, False]:
+      for isfloat16_beforesave in [True, False]:
+        for loaded_is_float16, load_dtype in [(False, dtypes.float32), (True, dtypes.float16), (isfloat16_beforesave, None)]:
+          torch_finally_isfloat16 = isfloat16_aftersave or isfloat16_beforesave
+          if torch_finally_isfloat16 == loaded_is_float16:
+            print(f"isfloat16_aftersave={isfloat16_aftersave}")
+            print(f"isfloat16_beforesave={isfloat16_beforesave}")
+            print(f"load_dtype={load_dtype}")
```

Comment:
so complicated and has redundant cases when you half model before and after

can just hardcode cases, would be more readable
```python
    cases = [
      # model is half, loaded dtype
      [1, dtypes.float16],
      [1, None],
      [0, dtypes.float32],
      [0, dtypes.float16],
      [0, None],
    ]
    
    # instead of
    # if isf16_after: model = model.half()
    if load_dtype == dtypes.float16: model = model.half()
```